doc/neps/nep-XXXX-SIMD-optimizations.rst

```diff
@@ -0,0 +1,199 @@
+===============================================================
+NEP XXXX — Using SIMD optimization instructions for performance
+===============================================================
+
+:Author: Sayed Adel, Matti Picus, Ralf Gommers
+:Status: Draft
+:Type: Standards
+:Created: 2019-11-25
+:Resolution: none
+
+
+Abstract
+--------
+
+While compilers are getting better at using hardware-specific routines to
+optimize code, they sometimes do not produce optimal results. Also, we would
+like to be able to copy binary C-extension modules from one machine to another
+without recompiling.
+
+We have a mechanism in the ufunc machinery to choose optimal loops. This NEP
+proposes a mechanism to build on that for many more features and architectures.
+The steps would be:
+
+- Establish a baseline of CPU features for minimal support
+- Write explicit code to take advantage of well-defined, architecture-agnostic,
+  universal intrisics which capture features available across architectures.
+- Capture those universal intrisics in a set of C macros that at compile time
+  would build code paths for each feature from the baseline up to the maximum
+  set of features available on that architecture.
```

Comment:
"each feature" is probably not going to be true. I would expect this to be a trade-off between number of features and binary size (and perhaps testability). For example, I'd expect us to build only the highest AVX512 variant, and not all of them.

Reply:
restated

Reply:
not pushed yet I think?